packages/flutter_tools/lib/src/run.dart

```diff
@@ -114,67 +116,114 @@ class RunAndStayResident extends ResidentRunner {
       }
     }
 
-    _package = getApplicationPackageForPlatform(device.platform, applicationBinary: applicationBinary);
+    _result = null;
+    Stopwatch startTime = new Stopwatch()..start();
 
-    if (_package == null) {
-      String message = 'No application found for ${device.platform}.';
-      String hint = getMissingPackageHintForPlatform(device.platform);
-      if (hint != null)
-        message += '\n$hint';
-      printError(message);
-      return 1;
+    String toolsExtension = Cache.instance.getToolsExtension('run');
+    if (toolsExtension != null) {
+      int observatoryPort;
+      const observatoryPrefix = 'Observatory listening on http://127.0.0.1:';
+      int diagnosticPort;
+      const diagnosticPrefix = 'Diagnostic server listening on http://127.0.0.1:';
+      Completer<bool> launched = new Completer<bool>();
+
+      await startEchoingDeviceLog();
+      int exitCode;
+      runCommandAndStreamOutput(
+        <String>[toolsExtension, _mainPath],
+        mapFunction: (String line) {
+          if (line.startsWith(observatoryPrefix))
+            observatoryPort = int.parse(line.substring(observatoryPrefix.length), onError: (_) => null);
+          if (line.startsWith(diagnosticPrefix))
+            diagnosticPort = int.parse(line.substring(diagnosticPrefix.length), onError: (_) => null);
+          if (!launched.isCompleted) {
+            if (line.contains('Application not launched by flutter-run script'))
```

Comment:
This should probably be a named constant and the expectation that the tools extension for 'run' must print this on launch failure has to be documented.